File: conformance/tests/ec-primitives/ec-primitives.runar.zig
const runar = @import("runar");

pub const ECPrimitives = struct {
    pub const Contract = runar.SmartContract;

    pt: runar.Point,

    pub fn init(pt: runar.Point) ECPrimitives {
        return .{ .pt = pt };
    }

    pub fn checkX(self: *const ECPrimitives, expectedX: i64) void {
        runar.assert(runar.ecPointX(self.pt) == expectedX);
    }

    pub fn checkY(self: *const ECPrimitives, expectedY: i64) void {
        runar.assert(runar.ecPointY(self.pt) == expectedY);
    }

    pub fn checkOnCurve(self: *const ECPrimitives) void {
        runar.assert(runar.ecOnCurve(self.pt));
    }

    pub fn checkNegateY(self: *const ECPrimitives, expectedNegY: i64) void {
        const negated = runar.ecNegate(self.pt);
        runar.assert(runar.ecPointY(negated) == expectedNegY);
    }

    pub fn checkModReduce(self: *const ECPrimitives, value: i64, modulus: i64, expected: i64) void {
        _ = self;
        runar.assert(runar.ecModReduce(value, modulus) == expected);
    }

    pub fn checkAdd(self: *const ECPrimitives, other: runar.Point, expectedX: i64, expectedY: i64) void {
        const result = runar.ecAdd(self.pt, other);
        runar.assert(runar.ecPointX(result) == expectedX);
        runar.assert(runar.ecPointY(result) == expectedY);
    }

    pub fn checkMul(self: *const ECPrimitives, scalar: i64, expectedX: i64, expectedY: i64) void {
        const result = runar.ecMul(self.pt, scalar);
        runar.assert(runar.ecPointX(result) == expectedX);
        runar.assert(runar.ecPointY(result) == expectedY);
    }

    pub fn checkMulGen(self: *const ECPrimitives, scalar: i64, expectedX: i64, expectedY: i64) void {
        _ = self;
        const result = runar.ecMulGen(scalar);
        runar.assert(runar.ecPointX(result) == expectedX);
        runar.assert(runar.ecPointY(result) == expectedY);
    }

    pub fn checkMakePoint(self: *const ECPrimitives, x: i64, y: i64, expectedX: i64, expectedY: i64) void {
        _ = self;
        const pt = runar.ecMakePoint(x, y);
        runar.assert(runar.ecPointX(pt) == expectedX);
        runar.assert(runar.ecPointY(pt) == expectedY);
    }

    pub fn checkEncodeCompressed(self: *const ECPrimitives, expected: runar.ByteString) void {
        const compressed = runar.ecEncodeCompressed(self.pt);
        runar.assert(compressed == expected);
    }

    pub fn checkMulIdentity(self: *const ECPrimitives) void {
        const result = runar.ecMul(self.pt, 1);
        runar.assert(runar.ecPointX(result) == runar.ecPointX(self.pt));
        runar.assert(runar.ecPointY(result) == runar.ecPointY(self.pt));
    }

    pub fn checkNegateRoundtrip(self: *const ECPrimitives) void {
        const neg1 = runar.ecNegate(self.pt);
        const neg2 = runar.ecNegate(neg1);
        runar.assert(runar.ecPointX(neg2) == runar.ecPointX(self.pt));
        runar.assert(runar.ecPointY(neg2) == runar.ecPointY(self.pt));
    }

    pub fn checkAddOnCurve(self: *const ECPrimitives, other: runar.Point) void {
        const result = runar.ecAdd(self.pt, other);
        runar.assert(runar.ecOnCurve(result));
    }

    pub fn checkMulGenOnCurve(self: *const ECPrimitives, scalar: i64) void {
        _ = self;
        const result = runar.ecMulGen(scalar);
        runar.assert(runar.ecOnCurve(result));
    }
};

File: conformance/tests/ec-primitives/ec-primitives.runar.go
//go:build ignore

package contract

import "runar"

type ECPrimitives struct {
	runar.SmartContract
	Pt runar.Point `runar:"readonly"`
}

func (c *ECPrimitives) CheckX(expectedX runar.Bigint) {
	runar.Assert(runar.EcPointX(c.Pt) == expectedX)
}

func (c *ECPrimitives) CheckY(expectedY runar.Bigint) {
	runar.Assert(runar.EcPointY(c.Pt) == expectedY)
}

func (c *ECPrimitives) CheckOnCurve() {
	runar.Assert(runar.EcOnCurve(c.Pt))
}

func (c *ECPrimitives) CheckNegateY(expectedNegY runar.Bigint) {
	negated := runar.EcNegate(c.Pt)
	runar.Assert(runar.EcPointY(negated) == expectedNegY)
}

func (c *ECPrimitives) CheckModReduce(value, modulus, expected runar.Bigint) {
	runar.Assert(runar.EcModReduce(value, modulus) == expected)
}

func (c *ECPrimitives) CheckAdd(other runar.Point, expectedX, expectedY runar.Bigint) {
	result := runar.EcAdd(c.Pt, other)
	runar.Assert(runar.EcPointX(result) == expectedX)
	runar.Assert(runar.EcPointY(result) == expectedY)
}

func (c *ECPrimitives) CheckMul(scalar, expectedX, expectedY runar.Bigint) {
	result := runar.EcMul(c.Pt, scalar)
	runar.Assert(runar.EcPointX(result) == expectedX)
	runar.Assert(runar.EcPointY(result) == expectedY)
}

func (c *ECPrimitives) CheckMulGen(scalar, expectedX, expectedY runar.Bigint) {
	result := runar.EcMulGen(scalar)
	runar.Assert(runar.EcPointX(result) == expectedX)
	runar.Assert(runar.EcPointY(result) == expectedY)
}

func (c *ECPrimitives) CheckMakePoint(x, y, expectedX, expectedY runar.Bigint) {
	pt := runar.EcMakePoint(x, y)
	runar.Assert(runar.EcPointX(pt) == expectedX)
	runar.Assert(runar.EcPointY(pt) == expectedY)
}

func (c *ECPrimitives) CheckEncodeCompressed(expected runar.ByteString) {
	compressed := runar.EcEncodeCompressed(c.Pt)
	runar.Assert(compressed == expected)
}

func (c *ECPrimitives) CheckMulIdentity() {
	result := runar.EcMul(c.Pt, 1)
	runar.Assert(runar.EcPointX(result) == runar.EcPointX(c.Pt))
	runar.Assert(runar.EcPointY(result) == runar.EcPointY(c.Pt))
}

func (c *ECPrimitives) CheckNegateRoundtrip() {
	neg1 := runar.EcNegate(c.Pt)
	neg2 := runar.EcNegate(neg1)
	runar.Assert(runar.EcPointX(neg2) == runar.EcPointX(c.Pt))
	runar.Assert(runar.EcPointY(neg2) == runar.EcPointY(c.Pt))
}

func (c *ECPrimitives) CheckAddOnCurve(other runar.Point) {
	result := runar.EcAdd(c.Pt, other)
	runar.Assert(runar.EcOnCurve(result))
}

func (c *ECPrimitives) CheckMulGenOnCurve(scalar runar.Bigint) {
	result := runar.EcMulGen(scalar)
	runar.Assert(runar.EcOnCurve(result))
}
